### PR TITLE
Database Structure Implementation

### PR DESCRIPTION
**Pull request by AlexSteeves**
Initial design of database, currently does not compile, I'm having a problem with foreign keys in the employees table referencing address info. The attribute values are also not correct I've set everything to VARCHAR during compilation.

### CODE REVIEW

File: InitialDatabaseSQL.sql [hunk 1]
```diff
@@ -0,0 +1,245 @@
+
+use Test;
+
+DROP TABLE address_info;
+
+CREATE TABLE address_info(
+	street_name VARCHAR(20) NOT NULL,
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:7]: Change to VARCHAR(30)

File: InitialDatabaseSQL.sql [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    street_number INT NOT NULL CHECK (street_number >= 0),
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:8]: Change to SMALLINT

File: InitialDatabaseSQL.sql [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    city VARCHAR(20) NOT NULL,
+    province VARCHAR(20) NOT NULL,
+    postal_code VARCHAR(6) NOT NULL,
+    country VARCHAR(20) NOT NULL,
+
+	PRIMARY KEY(street_name, street_number, postal_code)
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:14]: Remove the excess indent

File: InitialDatabaseSQL.sql [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+);
+
+DROP TABLE employee;
+
+CREATE TABLE employee(
+	emp_SSN VARCHAR(20) NOT NULL,
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:20]: remove the excess indent

File: InitialDatabaseSQL.sql [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    first_name VARCHAR(20) NOT NULL,
+    middle_name VARCHAR(20) NOT NULL,
+    last_name VARCHAR(20) NOT NULL,
+    
+    street_name VARCHAR(20) NOT NULL,
+    street_number INT NOT NULL CHECK (street_number >= 0),
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:26]: Change to SMALLINT

File: InitialDatabaseSQL.sql [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
+DROP TABLE employee;
+
+CREATE TABLE employee(
+	emp_SSN VARCHAR(20) NOT NULL,
+    first_name VARCHAR(20) NOT NULL,
+    middle_name VARCHAR(20) NOT NULL,
+    last_name VARCHAR(20) NOT NULL,
+    
+    street_name VARCHAR(20) NOT NULL,
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:25]: Change to VARCHAR(30)

File: InitialDatabaseSQL.sql [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    province VARCHAR(20) NOT NULL,
+    postal_code VARCHAR(6) NOT NULL,
+    country VARCHAR(20),
+    hotel_id VARCHAR(20),
+    start_date VARCHAR(20),
+    password VARCHAR(20),
+    last_update VARCHAR(20),
+    
+    Primary Key(emp_SSN),
+
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:37]: Use

`FOREIGN KEY (street_name, street_number, postal_code) REFERENCES address_info(street_name, street_number, postal_code)`

**Comment by MichiasShiferaw** [comment]
> Initial design of database, currently does not compile, I'm having a problem with foreign keys in the employees table referencing address info.

Should be solved with
`FOREIGN KEY (street_name, street_number, postal_code) REFERENCES address_info(street_name, street_number, postal_code)`

File: InitialDatabaseSQL.sql [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    middle_name VARCHAR(20) NOT NULL,
+    last_name VARCHAR(20) NOT NULL,
+    
+    street_name VARCHAR(20) NOT NULL,
+    street_number INT NOT NULL CHECK (street_number >= 0),
+    city VARCHAR(20) NOT NULL,
+    province VARCHAR(20) NOT NULL,
+    postal_code VARCHAR(6) NOT NULL,
+    country VARCHAR(20),
+    hotel_id VARCHAR(20),
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:31]: Add NOT NULL for remaining attributes

File: InitialDatabaseSQL.sql [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+);
+
+
+CREATE TABLE role(
+	emp_SSN VARCHAR(20) NOT NULL,
+    role_id VARCHAR(20) NOT NULL,
+    name VARCHAR(20) NOT NULL,
+    salary VARCHAR(20) NOT NULL,
+    
+    Primary Key(emp_SSN),
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:59]: Should be 
Primary Key(emp_SSN, role_id),

File: InitialDatabaseSQL.sql [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+    Primary Key(role_id),
+    
+    FOREIGN KEY(emp_SSN) REFERENCES employee(emp_SSN)
+);
+
+
+
+CREATE TABLE customer(
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:67]: make sure customer's datatypes are similar to employees

File: InitialDatabaseSQL.sql [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+    FOREIGN KEY(city) references address_info(city),
+    FOREIGN KEY(province) references address_info(province),
+    FOREIGN KEY(country) references address_info(country)
+
+
+);
+
+
+CREATE TABLE hotel_chain (
+	chain_id VARCHAR(20)  NOT NULL,
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:96]: CHAR(3)

File: InitialDatabaseSQL.sql [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+    FOREIGN KEY(postal_code) REFERENCES address_info(postal_code),
+    FOREIGN KEY(country) REFERENCES address_info(country)
+    
+
+);
+
+
+
+CREATE TABLE hotel (
+	chain_id VARCHAR(20)  NOT NULL,
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:125]: CHAR(3)

File: InitialDatabaseSQL.sql [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+    postal_code VARCHAR(20)  NOT NULL,
+    city VARCHAR(20)  NOT NULL,
+    province VARCHAR(20)  NOT NULL,
+    country VARCHAR(20)  NOT NULL,
+	manager_SSN VARCHAR(20)  NOT NULL,
+    num_of_rooms VARCHAR(20)  NOT NULL,
+    rating VARCHAR(20)  NOT NULL,
+    email VARCHAR(20)  NOT NULL,
+    
+    PRIMARY KEY(chain_id),
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:139]: Correct to
PRIMARY KEY(chain_id,hotel_id),

File: InitialDatabaseSQL.sql [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+    
+    FOREIGN KEY(manager_SSN) references employee(employee_SSN)
+);
+
+CREATE TABLE hotel_phone(
+	hotel_id VARCHAR(20)  NOT NULL,
+    phone_number VARCHAR(20)  NOT NULL,
+    department VARCHAR(20)  NOT NULL,
+    
+    PRIMARY KEY(hotel_id),
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:157]: PRIMARY KEY(chain_id,hotel_id),

File: InitialDatabaseSQL.sql [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+    hotel_id VARCHAR(20)  NOT NULL,
+    price VARCHAR(20)  NOT NULL,
+    room_category_id VARCHAR(20)  NOT NULL,
+    amenities VARCHAR(20)  NOT NULL,
+    damages VARCHAR(20)  NOT NULL,
+    last_updated VARCHAR(20)  NOT NULL,
+    
+    PRIMARY KEY(room_id),
+    
+    FOREIGN KEY(hotel_id) REFERENCES booking_info(hotel_id),
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:186]: REFERENCES hotel.hotel_id and not booking_info

File: InitialDatabaseSQL.sql [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+    emp_SSN VARCHAR(20)  NOT NULL,
+    arrival_date VARCHAR(20)  NOT NULL,
+    depature_date VARCHAR(20)  NOT NULL,
+    created_at VARCHAR(20)  NOT NULL,
+    last_updated VARCHAR(20)  NOT NULL,
+    
+    PRIMARY KEY(booking_id),
+    
+    FOREIGN KEY(customer_SSN) REFERENCES customer(customer_SSN),
+    FOREIGN KEY(employee_SSN) REFERENCES employee(employee_SSN)
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:206]: FOREIGN KEY(emp_SSN) REFERENCES employee(emp_SSN)

File: InitialDatabaseSQL.sql [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+    arrival_time VARCHAR(20)  NOT NULL,
+    created_at VARCHAR(20)  NOT NULL,
+    last_updated VARCHAR(20)  NOT NULL,
+    
+    PRIMARY KEY(renting_id),
+    
+    FOREIGN KEY(hotel_id) REFERENCES hotel(hotel_id),
+    FOREIGN KEY(emp_SSN) REFERENCES employee(emp_SNN),
+    FOREIGN KEY(booking_id) REFERENCES booking_info(booking_id),
+    FOREIGN KEY(customer_SSN) REFERENCES employee(emp_SSN),
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:232]: FOREIGN KEY(customer_SSN) REFERENCES customer(customer_SSN),

File: InitialDatabaseSQL.sql [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+    FOREIGN KEY(room_no) REFERENCES room(room_no)
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:233]: FOREIGN KEY(hotel_id,room_no) REFERENCES room(hotel_id,room_no)

File: InitialDatabaseSQL.sql [hunk 19]
```diff
@@ -0,0 +1,245 @@
+
+use Test;
+
+DROP TABLE address_info;
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:4]: For organization, have all the drop tables at the top as such,

drop table hotel_chain;
drop table hotel_phone;

drop table renting_info;
drop table booking_info;
drop table customer;
drop table role;

drop table room;
drop table hotel;
drop table employee;
drop table room_category;
drop table address_info;

File: InitialDatabaseSQL.sql [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+    country VARCHAR(20)  NOT NULL,
+	manager_SSN VARCHAR(20)  NOT NULL,
+    num_of_rooms VARCHAR(20)  NOT NULL,
+    rating VARCHAR(20)  NOT NULL,
+    email VARCHAR(20)  NOT NULL,
+    
+    PRIMARY KEY(chain_id),
+    PRIMARY KEY(hotel_id),
+    
+    FOREIGN KEY(street_name) references address_info(street_name),
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:142]: https://github.com/MichiasShiferaw/database-proj/pull/16/commits/b950a9ad4889d03c623a0c89fb996d5c1233aee7#r1136243720

File: InitialDatabaseSQL.sql (hunk 16 above)
Comment (MichiasShiferaw) [reply InitialDatabaseSQL.sql:206]: Also add
`FOREIGN KEY(hotel_id,room_no) REFERENCES room(hotel_id,room_no)`

File: InitialDatabaseSQL.sql [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+    booking_id VARCHAR(20)  NOT NULL,
+    has_booked VARCHAR(20)  NOT NULL,
+    departure_time VARCHAR(20)  NOT NULL,
+    arrival_time VARCHAR(20)  NOT NULL,
+    created_at VARCHAR(20)  NOT NULL,
+    last_updated VARCHAR(20)  NOT NULL,
+    
+    PRIMARY KEY(renting_id),
+    
+    FOREIGN KEY(hotel_id) REFERENCES hotel(hotel_id),
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:229]: `    FOREIGN KEY(hotel_id,room_no) REFERENCES room(hotel_id,room_no)`

Comment (MichiasShiferaw) [reply InitialDatabaseSQL.sql:229]: Remove that line for this one

File: InitialDatabaseSQL.sql [hunk 22]
```diff
@@ -0,0 +1,245 @@
...
+    FOREIGN KEY(emp_SSN) REFERENCES employee(emp_SNN),
```
Comment (MichiasShiferaw) [inline InitialDatabaseSQL.sql:230]: emp_SSN** not emp_SNN

**Review by kevinluong21 — CHANGES_REQUESTED** [review]
The chain_name, email, and province need to be longer than 20 (50 should work on the data I want to insert)

**Review by kevinluong21 — CHANGES_REQUESTED** [review]
- The chain_name, email, and province need to be longer than 20 (50 should work on the data I want to insert)
- if manager SSN must be not null when inserting hotel, it means that i have to insert a manager first, but hotel ID must be not null when inserting an employee, but you can't insert a hotel first. basically it's impossible to insert.

**Review by MichiasShiferaw — APPROVED** [review]

**Review by kevinluong21 — CHANGES_REQUESTED** [review]
Could the hotel_name attribute in hotel be extended to 100 characters?

**Comment by kevinluong21** [comment]
Can the prices for the rooms be changed to DECIMAL(6, 2) instead to accomodate for prices $100 and above?

**Pull request merged**